### ISSUE TITLE
Can max bitrate be improved with a gigabit ethernet adapter?

### ISSUE THREAD

**Issue by Raicuparta** [mariotaku/moonlight-tv#68]
It seems like you have the max bitrate set to 60Mbps. Is this due to the ethernet port on the TVs? I know some people use adapters [like these](https://www.amazon.com/dp/B00BBD7NFU?ref=ppx_pop_mob_ap_share) to get gigabit connections on their LG TVs. Do you think using something like this would allow for higher bitrates in Moonlight? I'm willing to test it myself, just want to check with you first to see if there's some other limitation I'm not aware of.

**Comment by mariotaku** [comment]
Bitrate caps are set for stability. Lag, corrupted frames or crash may occur on higher bitrate that isn't supported by webOS. See https://webostv.developer.lge.com/discover/specifications/supported-media-formats/webos-50/

**Comment by Raicuparta** [comment]
Got it, thanks.

**Comment by Raicuparta** [comment]
@mariotaku I was gonna try to build a version that allows for higher bitrate just to run some tests, but setting up the NDK seems like a bit of a doozy 😵 Any chance you could add some secret setting for disabling the bitrate cap? If not, I'll keep going with the setup process.

**Comment by roykaandorp** [comment]
@Raicuparta I was struggling with that too, but mariotaku made it really easy to do it yourself. The SDK you could download the precompiled one and install, git clone this repo, copy the easy script to the root of the repo directory and execute.

**Comment by Raicuparta** [comment]
@royka1 thanks for the tip! I'll give it a shot

**Comment by Raicuparta** [comment]
@royka1 I assume you did this in Linux? I tried in Windows via Linux subsystem but just get tons of missing dependencies :/

**Comment by mariotaku** [comment]
@Raicuparta Only a few dependencies are needed. Make sure you didn't miss any steps needed. If you want to get help about building moonlight, feel free to open a new issue.

**Comment by thehhmdb** [comment]
Looking at the webOS documents it seems that newer TVs support up to 100Mbps now. Can the maximum bitrate be increased? Maybe with an additional warning if needed?

**Comment by mariotaku** [comment]
> Looking at the webOS documents it seems that newer TVs support up to 100Mbps now. Can the maximum bitrate be increased? Maybe with an additional warning if needed? 

It looks like that's for 8K. Could you point me to the document?

**Comment by thehhmdb** [comment]
Yes, that's right it's for 8K (HEVC 7680x4320@60P Main/Main10@L6.1 Maximum **100Mbps):**
https://webostv.developer.lge.com/develop/specifications/video-audio-60

If a TV is powerful enough for 100Mbps at 8K would it be fine at 4K too? Or is it maybe a software limitation?

**Comment by mariotaku** [comment]
> Yes, that's right it's for 8K (HEVC 7680x4320@60P Main/Main10@L6.1 Maximum **100Mbps):** https://webostv.developer.lge.com/develop/specifications/video-audio-60
> 
> If a TV is powerful enough for 100Mbps at 8K would it be fine at 4K too? Or is it maybe a software limitation?

I think it's better not to assume such case, as anything beyond document stated would be undefined and unpredictable. On my previous TV (2019 4K model), it indeed didn't crash for bitrate over 60mpbs (for most of the time), but there's no apparent improvement either.

**Comment by mariotaku** [comment]
@Yahfz Not only the limit has been observed on real models, it's written on [LG developer's website](https://webostv.developer.lge.com/develop/specifications/video-audio-240). For the most recent models, the bitrate cap is still around 60mbps.

You're welcomed to modify the source and try on your device, also sharing your result would be helpful. If some machines can be confirmed to have higher actual limit, I can raise the bitrate limit for them specifically.

**Comment by mariotaku** [comment]
@Yahfz here you go: https://github.com/mariotaku/moonlight-tv/actions/runs/9932919706

You can make similar change by forking the repo. And it should automatically build for you.

**Comment by fehbari** [comment]
Found this thread while looking for the same info.

I tend to agree with @Yahfz that people installing this client on their TVs from developer mode should generally be savvy enough to handle an additional warning or checkbox about higher bitrates. There are already experimental settings in the app capable of breaking things (such as AV1, which doesn't work well at all for me). I don't see how bitrate is much different. It's also very easy to break your setup on the host side messing with Sunshine settings. This entire high-end, self-hosted streaming niche isn't exactly newbie friendly and I believe is more of an enthusiast's hobby. 😄

Kinda off-topic, but I wonder if LG has really been updating [that particular piece of documentation](https://webostv.developer.lge.com/develop/specifications/video-audio-240) for the capabilities of their latest chips. It has remained mostly the same since WebOS 6, and there's no mention of HFR, which we know their modern TVs support. This is all conjecture, of course, but it could be the case of outdated documentation, or lack of incentive to test beyond the limits of the typical streaming use case, for which 60 Mbps is usually enough.

All that said, I have a C2 and a C3 and I'll be happy to test that build on both and report back.

**Comment by fehbari** [comment]
@mariotaku Happy to report it does work well on the C3 at 150 Mbps (HEVC) using a USB to Ethernet adapter. I'll try the C2 later tonight. And also test without the adapter out of curiosity (since the TV's ethernet port is limited to 100 Mbps).

![20240716_212515.jpg](https://github.com/user-attachments/assets/fc86f1e1-1587-4403-945b-11cebbdffb77)

![20240716_212414.jpg](https://github.com/user-attachments/assets/956aedd3-890f-41cc-974b-5717859a0e39)

![20240716_212355.jpg](https://github.com/user-attachments/assets/9819c941-ba81-40bd-afaf-14bb90871992)

**Comment by fehbari** [comment]
> @fehbari How did you get this performance menu to show up? I've noticed that on iOS that option is available in the settings, it isn't there on my C1 however.
> 
> 

Try pressing RB/R1 + LB/L1 + View/Select + Menu/Start at the same time on your controller. It should display the menu.

**Comment by fehbari** [comment]
I have an interesting find on the C3 using the USB Ethernet adapter. At 150 Mbps H265, in certain scenes of certain games at _specific_ camera angles, there's a severe drop in performance and the _"Unstable connection"_ warning pops up. Move the camera away from those angles and performance improves.

Monitoring my local network during those _"Unstable connection"_ spikes, I can see the TV is receiving close to 30 MB/s (250 Mbps), which is the typical transfer limit of USB 2.0 ports this TV uses. Switching to 5 GHz Wi-Fi, the issue seems to go away, so it looks like it's more of a port bandwidth limitation than a processor one. Why LG went with 100 Mbps Ethernet and USB 2.0 ports on these $1k+ premium TVs is beyond me, but I digress...

![Captura de tela 2024-07-16 233349](https://github.com/user-attachments/assets/200deada-5cc3-40e3-8c49-62cf5b8bade4)

Interestingly, switching to AV1 the network transfers stick much closer to the 150 Mbps limit, and the issue doesn't happen. Unfortunately, AV1 has severe ghosting and stuttering issues at any bitrate for me, so I'm unable to use it.

![Captura de tela 2024-07-16 234310](https://github.com/user-attachments/assets/a0782d3b-8c6e-4fd9-8ac8-77a00cbdc992)

At 110 Mbps the issue goes away in my tests with the USB Ethernet adapter, so I believe that's close to the _wired_ upper limit for the C3 using H265. Still, that's a significant bump over the current limit of 65 Mbps.

**Performance drop at 150 Mbps:**

![20240716_225914.jpg](https://github.com/user-attachments/assets/4aa4384d-d3b6-430d-a359-f5b03bf08567)

**Normal performance at 150 Mbps:**

![20240716_230044.jpg](https://github.com/user-attachments/assets/46790890-033e-4c6f-b490-85a57892c229)

Back to the original request, personally I'm in favor of unlocking the bitrate given the results, even if that requires a new experimental toggle on top of the current warning - _"Unlock Bitrate (Experimental)"_ or something like that. If anything it future proofs the app for newer, more capable chips and ports.

Thinking about it, that'd also be in line with other Moonlight clients. For instance, I could run the Windows client on a "potato" laptop and set the highest bitrate @ 4K, then quickly find out it's a mistake and lower my settings. I do believe users are capable of figuring it out themselves and warnings should suffice.

I'll report the C2 results once I figure out a way around LG's limitation of one TV per dev account.

**Comment by fehbari** [comment]
Sorry for the double (triple?) post. I believe I found a fix for the performance issue I reported earlier when using the USB Ethernet adapter. The root of the issue is a significant amount of [bitrate overshoot when using H265 10 bit](https://github.com/LizardByte/Sunshine/issues/747#issuecomment-1381137204), exceeding the bandwidth of the USB 2.0 port the adapter is connected to.

Setting _Two-Pass Mode_ to _"Full Resolution"_ in _NVENC Encoder_ settings helps Sunshine better adhere to the bitrate limits and fixes the problem for me. Monitoring my network I see it stays much closer to 150 Mbps at all times, so it doesn't exceed the bandwidth of the USB Ethernet adapter.

![image](https://github.com/user-attachments/assets/b36866de-0c29-4eac-b284-def3031b4623)

The description for this setting says:

> Adds preliminary encoding pass. This allows to detect more motion vectors, better distribute bitrate across the frame and more strictly adhere to bitrate limits. Disabling it is not recommended since this can lead to occasional bitrate overshoot and subsequent packet loss.

So it makes sense it'd help here. I believe the equivalent AMD setting is _AMF Rate Control_ under _AMF Encoder_.

> This controls the rate control method to ensure we are not exceeding the client bitrate target. 'cqp' is not suitable for bitrate targeting, and other options besides 'vbr_latency' depend on HRD Enforcement to help constrain bitrate overflows.

Happy to say then 150 Mbps H265 10 bit is indeed viable on the C3 with an adapter, as long as your host is configured properly to prevent bitrate overshoot.
